Sequencing files associated with Chadarevian et al (2025) CSF1R-inhibitor resistant model for CNS-wide microglia replacement strategies  
The following data sets were used:  
  PMID: This Paper  
  GEO  
In this publication, Chadarevian et al RNA libraries were made using the Illumina TruSeq Stranded mRNA kit which utilizes oligo-dT beads to select and amplify mRNA. cDNA was synthesized using nine rounds of PCR and quality and concentrations of the DNA libraries were assayed using the Agilent 2100 Bioanalyzer high sensitivity DNA assay and the DNA high sensitivity Qubit. The libraries were quantified by Kapa qPCR, normalized to 1.5nM and then multiplexed at equimolar concentrations for sequencing on the Illumina NovaSeq 6000 platform. FASTQ files were then preprocessed using BBDuk79 to filter out ribosomal RNA and PhiX reads, trim Illumina adapters, and to trim any base pairs below a PHRED score of 28. The quality of sequencing files was then verified using FASTQC (version 0.11.9) and all files were determined to be of sufficient quality for downstream processing. Reads were then aligned to the murine GRCm39 (Ensembl release 107) transcriptome using Kallisto (version 0.46.1), transcripts were summarized to the gene level through tximport (version 1.14.2), and differential gene expression analysis was performed using DESeq2 (version 1.26.0) after removing genes with summed counts <10. 
